sapicore/engine/synapse/STDP.py

```diff
@@ -0,0 +1,161 @@
+        alpha_minus: float = 0.01,
+        **kwargs
+    ):
+        """Constructs an STDP synapse object connecting two ensembles to each other or an ensemble to itself.
+
+        :class:`~engine.synapse.Synapse` attributes should be given as keyword arguments by the calling method
+        unless default values are acceptable.
+        """
+        super().__init__(**kwargs)
+
+        # configurable attributes specific to STDP, over and above a static synapse.
+        self.tau_plus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + tau_plus
+        self.tau_minus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + tau_minus
+
+        self.mu_plus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + mu_plus
+        self.mu_minus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + mu_minus
+
+        self.alpha_plus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + alpha_plus
+        self.alpha_minus = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + alpha_minus
+
+        # state attributes specific to STDP synapses, over and above Synapse.
+        self.src_last_spiked = torch.zeros(self.matrix_shape[1], dtype=torch.float, device=self.device)
+        self.dst_last_spiked = torch.zeros(self.matrix_shape[0], dtype=torch.float, device=self.device)
+
+        # toggle learning switch (should be on during training and off during testing).
+        self.learning_switch = True
+
+    def update_weights(self) -> Tensor:
+        """STDP weight update implementation.
+
+        Returns
+        -------
+        Tensor
+            2D tensor containing weight differences (dW).
+
+        """
+        # update last spike time to current simulation step iff a spike has just arrived at the synapse.
+        self.src_last_spiked = self.src_last_spiked + self.delayed_data * (-self.src_last_spiked + self.simulation_step)
+
+        # update last spike time only for postsynaptic units that spiked in this model iteration.
+        self.dst_last_spiked = self.dst_last_spiked + self.dst_ensemble.spiked * (
+            -self.dst_last_spiked + self.simulation_step
+        )
+
+        # initialize dW matrix--remember the format is dst.num_units by src.num_units.
+        delta_weight = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device)
+
+        # subtract postsynaptic last spike time stamps from presynaptic.
+        # transpose dst_last_spiked to a column vector, extend column-wise, then add its negative to src_last_spiked.
+        dst_tr = self.dst_last_spiked.reshape(self.dst_last_spiked.shape[0], 1)
+        delta_spike = -dst_tr.repeat(dst_tr.shape[1], self.src_ensemble.num_units) + self.src_last_spiked
+
+        # spike time differences for the potentiation and depression cases (pre < post, pre > post, respectively).
+        ltp_diffs = (delta_spike < 0.0).int() * delta_spike
```

Comment:
I think there are ways to do these ops in place and using some of torch's functions that are more efficient. But that can be left for later.

Reply:
Absolutely. After the merger, let's start an optimization branch.

Reply:
Ok, but please 🙏🙏🙏 no overall optimization branch, but specific optimization PRs 😅